### ISSUE TITLE
Regression in 4.1-rc-1 when generating IDEA module due to multiple attempts to resolve the same configuration

### ISSUE THREAD

**Issue by wilkinsona** [gradle/gradle#2573]
This was originally reported in a [comment](https://github.com/spring-gradle-plugins/dependency-management-plugin/issues/181#issuecomment-317334884) on a [Dependency Management Plugin](https://github.com/spring-gradle-plugins/dependency-management-plugin) issue.  I am raising it here as it appears to be regression in 4.1-rc-1.

A combination of the dependency management plugin configuring exclusions and a mixture of compile and compile only dependencies causes the IDEA plugin to attempt to resolve the `compileClasspath` configuration twice. This failure occurs with 4.1-rc-1. It does not occur with 4.0.1.

### Expected Behavior

The `ideaModule` task completes successfully

### Current Behavior

The `ideaModule` task fails:

```
./gradlew ideaModule --stacktrace

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':ideaModule'.
> Attempted to resolve configuration ':compileClasspath' that has been resolved previously.

* Try:
Run with --info or --debug option to get more log output.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':ideaModule'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:100)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:70)
        at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:63)
        at org.gradle.api.internal.tasks.execution.ResolveTaskOutputCachingStateExecuter.execute(ResolveTaskOutputCachingStateExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:58)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:88)
        at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker$1.run(DefaultTaskGraphExecuter.java:248)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:197)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:107)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:241)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:230)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.processTask(DefaultTaskPlanExecutor.java:124)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.access$200(DefaultTaskPlanExecutor.java:80)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:105)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:99)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.execute(DefaultTaskExecutionPlan.java:625)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.executeWithTask(DefaultTaskExecutionPlan.java:580)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.run(DefaultTaskPlanExecutor.java:99)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
Caused by: org.gradle.api.InvalidUserDataException: Attempted to resolve configuration ':compileClasspath' that has been resolved previously.
        at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.resolveGraphIfRequired(DefaultConfiguration.java:448)
        at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.resolveToStateOrLater(DefaultConfiguration.java:437)
        at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration.access$1600(DefaultConfiguration.java:114)
        at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$ConfigurationResolvableDependencies.getResolutionResult(DefaultConfiguration.java:1022)
        at org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$ConfigurationResolvableDependencies_Decorated.getResolutionResult(Unknown Source)
        at org.gradle.plugins.ide.internal.resolver.DefaultIdeDependencyResolver.getIncomingResolutionResult(DefaultIdeDependencyResolver.java:210)
        at org.gradle.plugins.ide.internal.resolver.DefaultIdeDependencyResolver.getIdeProjectDependencies(DefaultIdeDependencyResolver.java:57)
        at org.gradle.plugins.ide.internal.IdeDependenciesExtractor.extractProjectDependencies(IdeDependenciesExtractor.java:60)
        at org.gradle.plugins.ide.idea.model.internal.IdeaDependenciesProvider.getProjectDependencies(IdeaDependenciesProvider.java:93)
        at org.gradle.plugins.ide.idea.model.internal.IdeaDependenciesProvider.getDependencies(IdeaDependenciesProvider.java:82)
        at org.gradle.plugins.ide.idea.model.internal.IdeaDependenciesProvider.provide(IdeaDependenciesProvider.java:59)
        at org.gradle.plugins.ide.idea.model.IdeaModule.resolveDependencies(IdeaModule.java:535)
        at org.gradle.plugins.ide.idea.model.IdeaModule.mergeXmlModule(IdeaModule.java:548)
        at org.gradle.plugins.ide.idea.GenerateIdeaModule.configure(GenerateIdeaModule.java:43)
        at org.gradle.plugins.ide.idea.GenerateIdeaModule.configure(GenerateIdeaModule.java:32)
        at org.gradle.plugins.ide.api.XmlGeneratorTask$1.configure(XmlGeneratorTask.java:40)
        at org.gradle.plugins.ide.api.XmlGeneratorTask$1.configure(XmlGeneratorTask.java:32)
        at org.gradle.plugins.ide.api.GeneratorTask.generate(GeneratorTask.java:81)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:73)
        at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.doExecute(DefaultTaskClassInfoStore.java:141)
        at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.execute(DefaultTaskClassInfoStore.java:134)
        at org.gradle.api.internal.project.taskfactory.DefaultTaskClassInfoStore$StandardTaskAction.execute(DefaultTaskClassInfoStore.java:121)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:731)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:705)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$1.run(ExecuteActionsTaskExecuter.java:122)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:197)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:107)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:111)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:92)
        ... 27 more


* Get more help at https://help.gradle.org

BUILD FAILED in 0s
1 actionable task: 1 executed
```

When the failure occurs, resolving of `compileClasspath` appears to be triggered from lines 534 and 535 of `org.gradle.plugins.ide.idea.model.IdeaModule`.

Thus far, I've been unable to reproduce the problem without the dependency management plugin. I'll update this issue if I manage to do so.

### Steps to Reproduce (for bugs)

The problem can be reproduced by running the `ideaModule` task using 4.1-rc-1 with this `build.gradle`:

```groovy
plugins {
    id 'io.spring.dependency-management' version '1.0.3.RELEASE'
    id 'java'
    id 'idea'
}

dependencyManagement {
    dependencies {
        dependency('io.dropwizard.metrics:metrics-core:3.1.0') {
            exclude 'org.slf4j:slf4j-api'
        }
    }
}

repositories {
    mavenCentral()
}

dependencies {
    compileOnly 'org.slf4j:slf4j-api:1.7.12'
    compile 'io.dropwizard.metrics:metrics-core'
}
```

### Your Environment

This problem doesn't appear to be environment specific. I'm running:

```
------------------------------------------------------------
Gradle 4.1-rc-1
------------------------------------------------------------

Build time:   2017-07-20 15:22:15 UTC
Revision:     e9f4cb3be4b0a16506ad54b44485b7f0862f6c59

Groovy:       2.4.11
Ant:          Apache Ant(TM) version 1.9.6 compiled on June 29 2015
JVM:          1.8.0_141 (Oracle Corporation 25.141-b15)
OS:           Mac OS X 10.12.4 x86_64
```

**Comment by jjohannes** [comment]
Thanks for letting us know @wilkinsona. It looks as if the fix for the dependency exclusion issue actually introduced this. The problem is not in `4.0.1`, but in our `4.0.2` branch. 

- https://github.com/gradle/gradle/issues/2480
- https://github.com/gradle/gradle/commit/6a28fadabeb7b3cacf65a1ea6e4fa6534e5af673

We need to take a closer look at this @bigdaz.

**Comment by bigdaz** [comment]
Fixed in https://github.com/gradle/gradle/commit/56bf8b28ed021380e2da31b7372effc0521aee21

**Comment by eriwen** [comment]
@wilkinsona It'd be very helpful if you would verify the fix in Gradle versions `4.0.2-20170725142320+0000` and `4.1-20170725144305+0000`.

**Comment by wilkinsona** [comment]
@eriwen [Done](https://github.com/spring-gradle-plugins/dependency-management-plugin/issues/181#issuecomment-317849769)
